File: cavity_detection_api/src/cavity_detection_api/test_client.cpp
#include <ros/ros.h>
#include "geometry_msgs/Pose.h"
#include "cavity_detection_api/api.h"
#include <string>

int main(int argc, char** argv)
{
    ros::init(argc, argv, "cavity_client");

    geometry_msgs::Pose robot_pose;
    robot_pose.position.x = 1.0;
    robot_pose.position.y = 2.0;
    robot_pose.position.z = 3.0;

    geometry_msgs::Pose roi_pose;
    std::string roi_id;
    getNearestCavity(robot_pose, roi_id, roi_pose);

    roi_pose.position.x = 4.0;
    roi_pose.position.y = 5.0;
    roi_pose.position.z = 6.0;

    updateRoi(roi_id, roi_pose, 10.0, 5.0, 2.0);

    std::string cavity_id;
    addCavity(roi_id, 2.0, 3.0, cavity_id);

    updateCavity(roi_id, cavity_id, 2.0, 3.0, 0);

    return 0;
}

File: cavity_detection_api/scripts/cavity_detection_api/test_client.py
#!/usr/bin/env python3

import rospy
from cavity_detection_msgs.srv import *
from cavity_detection_api.api import get_nearest_cavity, update_roi, add_cavity, update_cavity
from geometry_msgs.msg import Pose, Point, Quaternion

def main():
    rospy.init_node('cavity_client_node')

    # Test get_nearest_cavity service
    roi_id, roi_pose = get_nearest_cavity()

    # Example ROI pose for testing
    new_roi_pose = Pose()
    new_roi_pose.position.x = 1.0
    new_roi_pose.position.y = 2.0
    new_roi_pose.position.z = 0.0
    new_roi_pose.orientation.x = 0.0
    new_roi_pose.orientation.y = 0.0
    new_roi_pose.orientation.z = 0.0
    new_roi_pose.orientation.w = 1.0

    # Test update_roi service
    update_roi(roi_id=roi_id, roi_pose=roi_pose, length=1.0, width=1.0, depth=1.0)

    # Test add_cavity service
    cavity_id = add_cavity(roi_id=roi_id, y_offset=0.5, width=0.2)
    print(cavity_id)
    # Test update_cavity service
    update_cavity(roi_id=roi_id, cavity_id=cavity_id, y_offset=3.5, width=0.5, status=True)

if __name__ == '__main__':
    main()